「コメントを付けられるようにする」の課題を作成しました

a-kuroki-gs opened the pull request.

a-kuroki-gs committed:
ReportモデルとCommentモデルの作成と紐づけを行いました

a-kuroki-gs committed:
ヘッダーに日報へのリンクを追加

a-kuroki-gs committed:
routes.rbに resources :reports を追加

a-kuroki-gs committed:
日報関連のコントローラの作成

a-kuroki-gs committed:
日報関連のビューの作成

app/view/booksの内容を参考にして作成

a-kuroki-gs committed:
ビューの細かい修正を行う

作成日は日本時間で、日まで表示
Contentに改行が反映されるようにする

a-kuroki-gs committed:
自分の日報のみ編集・削除できるようにする

作成者の情報を追加し、deviseのときと同じようにリンクを編集

a-kuroki-gs committed:
日報にコメントをつける機能を実装しました

a-kuroki-gs committed:
不要な空白の削除

a-kuroki-gs committed:
dependent: :destroyを追記

UserやReport・Bookに紐づいてデータが削除されるようにするため

a-kuroki-gs committed:
日報のコメント削除機能を実装しました

a-kuroki-gs committed:
コメント作成、表示の機能を、本と日報で共通化しました

a-kuroki-gs committed:
本の編集画面が正しく表示されるように、メソッドを指定しました

a-kuroki-gs committed:
不要なスペースの削除

a-kuroki-gs committed:
コメント作成者の表記を変更

作成者のタイトルを追加
名前が未登録であればメールアドレスを表示する

a-kuroki-gs committed:
JSONリクエストの部分を削除

a-kuroki-gs committed:
日報の編集画面が正しく表示されるように、メソッドを指定しました

a-kuroki-gs committed:
日本語化対応を行いました

a-kuroki-gs committed:
不要な記述の削除

コメント編集は実装しないことにしたため

a-kuroki-gs committed:
h2の見出しの大きさを変更しました

a-kuroki-gs committed:
マイグレーションファイルとschema.rbをまとめてコミットする

yuuu commented on app/controllers/books_controller.rb:
```diff
@@ -9,7 +9,10 @@ def index
-  def show; end
+  def show
+    @comment = Comment.new
+    @comments = @book.comments
```
`comments` も念のため `order` しておきましょう。DB次第で順番がランダムになる可能性があるので。

a-kuroki-gs replied:
`order`で並べるように変更しました。

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -0,0 +1,9 @@
+
+class Books::CommentsController < CommentsController
+  def create
+    @imageable = Book.find(params[:book_id])
```
`imageable` だと `イメージできる` という意味になってしまうので、別の名前を検討しましょう 🙏

a-kuroki-gs replied:
`commentable`に変更しました。

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -0,0 +1,37 @@
+  def create
+    @comment = @imageable.comments.build(comment_params)
+    @comment.user_id = current_user.id
+    @comment.imageable_id = @imageable.id
```
`@imageable.comments.build` としているのでこの行は不要なはずです。

a-kuroki-gs replied:
`build`によってID(今回は親のID)が自動で結びつく、ということが抜けていました。
不要な文を削除しました。

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -0,0 +1,37 @@
+    @comment.destroy
+
+    respond_to do |format|
+      format.html { redirect_to report_url(@comment.imageable), notice: t('controllers.common.notice_destroy', name: Comment.model_name.human) }
```
ここは `report_url` で問題ないでしょうか？ `book_url` の可能性はありませんか？

a-kuroki-gs replied:
どちらでも対応できるようにするのが抜けておりました。
`polymorphic_path`を利用して親へのリンクになるように変更しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,62 @@
+
+class ReportsController < ApplicationController
+  def index
+    @reports = Report.all
```
ここも明示的に `order` してできればページネーションも対応しましょう

a-kuroki-gs replied:
`order`とページネーションを実装しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,62 @@
+  end
+
+  def create
+    @report = Report.new(report_params)
```
ここは `current_user.reports.build(report_params)` とすると、自動的に `user_id` がセットされるので、次の行を省略できますね 👍

a-kuroki-gs replied:
`build`を用いてユーザーのIDと自動的に紐づけるようにしました。
また、モデルで定義できていなかったので、日報とユーザーの関連付けも行いました。

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -0,0 +1,37 @@
+  private
+
+  def comment_params
+    params.require(:comment).permit(:content, :user_id, :imageable_id)
```
`imageable_id` はフォームから入力されないと思いますが、ここに必要でしょうか？

a-kuroki-gs replied:
必要なかったので削除しました。

yuuu commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <p>
+    <strong><%= Comment.human_attribute_name(:creator) %>:</strong>
+    <% user = User.find(comment.user_id) %>
+    <%= user.name.presence || user.email %>
```
この処理は `User` クラス内にメソッドとして定義するとViewが見やすくなりますし、他のViewでも使い回せるようになります。

a-kuroki-gs replied:
`user.rb`内にメソッドとして定義し、ビューに記載する内容をコメント・日報から減らしました。

yuuu commented on app/views/reports/_form.html.erb:
```diff
@@ -0,0 +1,41 @@
+  </div>
+
+  <div>
+    <% unless controller.action_name == 'new' %>
```
これはフォームからPOSTせずとも、Reportをcreateした時点で自動的にセットされますね。次の `creator` も同様です。 `creator` に関してはstrong parameterで許可されていないので、おそらく今も機能していないかと

a-kuroki-gs replied:
日報入力フォームの作成者・作成日時の部分は不要だと判断したので、削除しました。

yuuu commented on db/migrate/20231024050342_add_imageable_to_comments.rb:
```diff
@@ -0,0 +1,6 @@
+class AddImageableIdToComments < ActiveRecord::Migration[7.0]
+  def change
+    add_column :comments, :imageable_id, :bigint
```
この書き方だとIndexが貼られていないですね。
ポリモーフィックのためのカラムは、専用の書き方が用意されているのでそちらを使うようにすればIndexの貼り忘れを防げます。
https://railsguides.jp/association_basics.html#%E3%83%9D%E3%83%AA%E3%83%A2%E3%83%BC%E3%83%95%E3%82%A3%E3%83%83%E3%82%AF%E9%96%A2%E9%80%A3%E4%BB%98%E3%81%91

a-kuroki-gs replied:
リンクに沿って、Indexを貼ることができるようにしました。

yuuu commented on db/migrate/20231024094835_add_imageable_id_to_comments.rb:
```diff
@@ -0,0 +1,6 @@
+class AddImageableIdToComments < ActiveRecord::Migration[7.0]
+  def change
+    add_reference :comments, :report, foreign_key: true
```
これらのカラムは不要なはずです。

a-kuroki-gs replied:
不要でしたので削除しました。

yuuu commented on config/routes.rb:
```diff
@@ -2,6 +2,13 @@
   resources :users, only: %i(index show)
+
+  resources :reports do
+    resources :comments, module: 'reports'
```
`CommentsController` に存在するアクションは限られているはずなので、 `only` オプションを使用して許可するアクションを絞り込むようにしましょう。

a-kuroki-gs replied:
`only`オプションを使用して許可するアクションを絞り込みました。
また、`reports`・`books`によらないものはまとめることにしたため、ビューも合わせて変更しました。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -0,0 +1,62 @@
+    end
+  end
+
+  def update
```
View側で自分の日報のみ編集や削除ボタンを表示するようにしているものの、HTMLはユーザー自身で改変ができるので、同じようなフォームを組み立てられてしまうと自分以外の日報でも操作できてしまう可能性があります。
別のユーザーがリクエストしてきた時にそれを拒否するロジックを追加しましょう。

a-kuroki-gs replied:
自分以外の日報・コメントの編集・削除ができないロジックを実装しました。

yuuu replied:
コメントと同様にここも `current_user.reports.find(params[:id])` とするとif式を省略できますね。 `destroy` も同様です。

a-kuroki-gs committed:
imageable(親)をcommentableに変更

a-kuroki-gs committed:
commentもIDで並べるようにする

a-kuroki-gs committed:
日報一覧にもページネーションを対応させる

a-kuroki-gs committed:
親のIDを紐づけしていた部分が不要だったため削除

buildにより親のIDは自動で紐づくため

a-kuroki-gs committed:
Strong Parameterの不要なカラムを削除

a-kuroki-gs committed:
buildを使ってユーザーのIDと自動的に紐づける

モデルで定義できていなかったので、日報とユーザーの関連付けも行う

a-kuroki-gs committed:
onlyオプションを使用して許可するアクションを絞り込む

reports・booksによらないものはまとめることにしたため、ビューも合わせて変更

a-kuroki-gs committed:
リダイレクト先の変更

コメント作成・削除後、日報へのリンクとなっていたので、polymorphic_pathを利用して親へのリンクになるように変更

a-kuroki-gs committed:
ビューで処理していた内容をUserクラスにメソッドとして定義

ビューに記載する内容をコメント・日報から減らす

a-kuroki-gs committed:
不要なカラムの消去

a-kuroki-gs committed:
Indexを貼ることができるようにする

a-kuroki-gs committed:
うまく表示できていなかった日報入力フォームの作成者・作成日時の部分を削除

a-kuroki-gs committed:
自分以外の日報・コメントの編集・削除ができないロジックを実装

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  def destroy
+    @comment = Comment.find(params[:id])
```
ここは `current_user.comments.find(params[:id])` とすれば、自分以外のコメントを指定されたときに404を返すことができ、if式が不要になりますね。

a-kuroki-gs replied:
日報作成時などでは`current_user`を利用して作成していたにも関わらず、
そのロジックは思いつきませんでした。。。
大変勉強になりました、ありがとうございます！

日報の編集・削除、コメントの削除について修正しております。

a-kuroki-gs committed:
日報の編集・削除、コメントの削除において、current_userを基準にして日報・コメントを探すように修正

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -0,0 +1,36 @@
+      if @comment.save
+        format.html { redirect_to polymorphic_path(@comment.commentable), notice: t('controllers.common.notice_create', name: Comment.model_name.human) }
+      else
+        format.html { redirect_to polymorphic_path(@comment.commentable), status: :unprocessable_entity }
```
前回コメントできておらずすみません 🙏 

ここで `redirect` をしてしまうと、バリデーションエラー時にそのフォームに入力していた内容が全て消えてしまい、ユーザーとしてはストレスを感じてしまいます。(今回はコメントの本文のみがバリデーション対象なので特に違和感ないですが他の入力欄を増やそうとした時にこの問題が顕在化します)

scaffoldで作ったコントローラ通りにここは `render` するのが望ましいのでそのように修正いただきたいです 🙏

a-kuroki-gs replied:
ありがとうございます！
課題の方でもコメントさせていただきましたが、こちら`render`に修正し、内容が消えないようにいたしました。

a-kuroki-gs committed:
コメントの保存に失敗した時、redirectしていたのをrenderするように修正

render先をReportとBookで分けて指定するためのメソッドをそれぞれ定義
それぞれのshow.html.erbにおいて、日報表示に必要なrender先をreports/reportとディレクトリから明記
コメント保存に失敗した時のエラーメッセージを表示していなかったので追加

yuuu commented on app/controllers/concerns/comments_controller.rb:
```diff
@@ -13,7 +13,9 @@ def create
-        format.html { redirect_to polymorphic_path(@comment.commentable), status: :unprocessable_entity }
+        transform_commentable
+        @comments = @commentable.comments.order(:id)
+        format.html { render "#{@comment.commentable_type.downcase}s/show", status: :unprocessable_entity }
```
確かにこの方法で今の使用であれば動作しますが英語の複数形は単純に `s` を付けるだけとは限らないので、例えば `History` のようなモデル名が出てきた時に英語として正しくなくなる懸念がありますね。

ここは「コメント対象によって変えたい部分」なので、このコントローラでなんとかするのではなく `book/comments_controller` と `report/comments_controller` を指定する形にしてはどうでしょうか？

----
ちなみにRails(というよりActiveSupport)のメソッドで、英語的に正しく複数形にするメソッドとして [pluralize](https://railsdoc.com/page/pluralize) というものが用意されています。今回は使う必要はないと思いますが参考まで。

a-kuroki-gs replied:
共通化を意識しすぎて、かえってわかりにくいコードになってしまいました。

コントローラ別で切り分けてメソッドを作成し、
それぞれのコントローラ内のメソッドで処理するように変更しました。

a-kuroki-gs committed:
コメント保存失敗時のrender先を、コントローラ別に作成しました

yuuu commented on app/controllers/books/comments_controller.rb:
```diff
@@ -9,7 +9,9 @@ def create
-  def transform_commentable
+  def render_commentable_show(format)
     @book = @commentable
+    @comments = @book.comments.order(:id)
```
今のコードは共通な部分まで、子クラス側に持って行ってしまっていて `render_commentable_show` を呼ぶためには呼び元で `respond_to` して `format` を渡さなければならないという問題がありますね。

BookとReportで挙動を変えたいのは `render` の引数だけのはずなので、ここだけを子クラスに持っていけないか検討してみてください 🙏

a-kuroki-gs replied:
`@report`や`@book`の設定もしないといけないから…とまとめてしまっておりました。
`render`先のみ子クラスで明記し、`@report`や`@book`の設定は子クラスの`create`メソッドで行うように修正しました。

a-kuroki-gs committed:
render先のみそれぞれのコントローラで明記しました

共通な部分も子クラス側に持って行ってしまっていたため
